File: Leetcode/Python3/649. Dota2 Senate/solution.py
# https://leetcode.com/problems/dota2-senate/description/


from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r = d = 0
        queue = deque()
        for team in senate:  # Append to queue and count each team
            queue.append(team)
            if team == "R":
                r += 1
            else:
                d += 1

        ban_r = ban_d = 0
        while r != 0 and d != 0:
            cur = queue.popleft()
            if cur == "R":
                if ban_r == 0:  # If no ban Radiant, Radiant play next round and ban Dire
                    queue.append(cur)
                    ban_d += 1
                else:   # Dire ban Radiant so cant exercise
                    ban_r -= 1
                    r -= 1
            elif cur == "D":
                if ban_d == 0:  # If no ban Dire, Dire play next round and ban Radiant
                    queue.append(cur)
                    ban_r += 1
                else:  # Radiant ban Dire so cant exercise
                    ban_d -= 1
                    d -= 1
        return "Radiant" if d == 0 else "Dire"

File: Leetcode/C++/649. Dota2 Senate/solution.cpp
// https://leetcode.com/problems/dota2-senate/description/

#include <queue>
#include <string>

using namespace std;

class Solution
{
public:
    string predictPartyVictory(string senate)
    {
        queue<char> q;
        int r = 0, d = 0;
        for (int i = 0; i < senate.size(); i++) // Append to queue and count each team
        {
            q.push(senate[i]);
            if (senate[i] == 'R')
            {
                r += 1;
            }
            else
            {
                d += 1;
            }
        }
        int banR = 0, banD = 0;
        while (r != 0 && d != 0)
        {
            char cur = q.front();
            q.pop();
            if (cur == 'R')
            {
                if (banR > 0) // Dire ban Radiant so cant exercise
                {
                    banR -= 1;
                    r -= 1;
                }
                else // If no ban Radiant, Radiant play next round and ban Dire
                {
                    q.push(cur);
                    banD += 1;
                }
            }
            else
            {
                if (banD > 0) // Radiant ban Dire so cant exercise
                {
                    banD -= 1;
                    d -= 1;
                }
                else // If no ban Dire, Dire play next round and ban Radiant
                {
                    q.push(cur);
                    banR += 1;
                }
            }
        }
        return r == 0 ? "Dire" : "Radiant";
    }
};